Nix packaging

lbeierlieb opened the pull request:
This PR adds Nix packaging for smartvmi and all smartvmi's dependencies that are not yet packaged in nixpkgs.
The Nix-based build locks down all dependencies and prevents online access during the build process, resulting in reproducible builds and avoiding build behavior that is based on the configuration of the system the build runs on.
NixOS is not required to utilize smartvmi built by Nix—installing the Nix package manager, e.g., on Debian, is sufficient.

lbeierlieb committed:
cmake: allow libvmi to be provided externally

the default behavior (retrieving libvmi from
github:GDATASoftwareAG/libvmi) is not changed, but this change allows
vcpkg or Nix to supply libvmi instead.

lbeierlieb committed:
cmake: allow corrosion to be provided externally

the default behavior (retrieving Corrosion from
github:corrosion-rs/corrosion) is not changed, but this change allows
vcpkg or Nix to supply Corrosion instead.

phip1611 commented on nix/bext-di.nix:
```diff
@@ -0,0 +1,24 @@
+  version = "v1.3.2";
+
+  src = builtins.fetchTarball {
+    url = "https://github.com/boost-ext/di/archive/refs/tags/${version}.tar.gz";
```
why isn't this a flake input? We can keep it that way, but when we already use flakes..

lbeierlieb replied:
I implemented the fetch this way because I saw the same pattern many times in derivations in nix packages. I have only used flake inputs to retrieve other flakes, so the idea did not even arise.
I like the suggestion but am unsure about some details. With the current implementation, I can update the version, and  the tarball URL is automatically adapted. With flake inputs, would you recommend specifying a branch or a tag? My considerations and questions:

- branch (e.g., main, stable)
  - flake update also updates to newest revision on the branch, can update to newer version
- tag (e.g., v1.3.2)
  - flake input always points to same version, perhaps always to same commit
  - version upgrade requires specifying another tag

In both cases, how do you handle the version field in the derivation? If you specified a branch, the version might change on flake update, but you have to manually check this and update the version field in the derivation. With a tag, you have to be aware of new releases and update flake input and version field. I don't like that this requires changes in two separate files (flake.nix, bext-di.nix), and I found that parsing the version from the flake input URL is possible. Is this approach good practice according to your experience?

snue replied:
> In both cases, how do you handle the version field in the derivation?

As you rely on a versioned tarball currently, you would point to a stable tag. You are right, this requires keeping the version in the input and the derivation in sync on updates.  You can probably build convenience around that, but I have commonly seen it done manually before.

phip1611 commented on nix/libvmi-gdata.nix:
```diff
@@ -0,0 +1,53 @@
+    cmake
+    pkg-config
+    bison
+    flex
```
nit: not alphabetically ordered

phip1611 commented on nix/libvmi-gdata.nix:
```diff
@@ -0,0 +1,53 @@
+  ];
+  buildInputs = [
+    json_c
+    glib
```
nit: not alphabetically ordered

phip1611 commented on vmicore/rust_src/Cargo.toml:
```diff
@@ -12,8 +12,8 @@ prost = "0.13.3"
-hyper-util = "0.1.10"
-cxx = "1.0"
+hyper-util="0.1.10"
+cxx = "=1.0.129"# must match version on 'cxx'st match version on 'cxxbridge-cmd'
```
the comment sounds a little broken? two times "match"?

phip1611 commented on vmicore/rust_src/Cargo.toml:
```diff
@@ -26,4 +26,5 @@ tower = "0.5.1"
 ctrlc = "3.4.5"
 
 [build-dependencies]
+cxxbridge-cmd = "=1.0.129" # must match version on 'cxx'
```
on => of?

phip1611 commented on nix/build_vmicore.nix:
```diff
@@ -0,0 +1,84 @@
+      rustc
+      tclap
+      zlib
+      corrosion
```
not alphabetical

phip1611 commented on nix/build_vmicore.nix:
```diff
@@ -0,0 +1,84 @@
+      tclap
+      zlib
+      corrosion
+      libvmi
```
isn't libvmi just a buildInput?

lbeierlieb replied:
yes! there were actually quite a few packages listed in the incorrect dependency category. Fixed

snue commented on vmicore/Readme.md:
```diff
@@ -4,6 +4,7 @@
 # VMICore
 
 _VMICore_ is a VM Introsepction tool capable of dynamic malware analysis.
+If you prefer to build with Nix, see [the Nix Readme](../nix/Readme.md).
```
This remark should move under the "How to Build" headline, maybe as a note/tip at the top (see https://github.blog/changelog/2023-12-14-new-markdown-extension-alerts-provide-distinctive-styling-for-significant-content/).

```suggestion
[!TIP]
If you prefer to build with Nix, see [the Nix Readme](../nix/Readme.md).
```

snue commented on vmicore/Readme.md:
```diff
@@ -7,6 +7,9 @@ _VMICore_ is a VM Introsepction tool capable of dynamic malware analysis.
 
 ## How to Build
 
+[!NOTE]
+If you prefer to build with Nix, see [the Nix Readme](../nix/Readme.md).
```
From what this does, I think `[!TIP]` is a better semantic fit. But a note is fine as well.

> [!TIP]
> Helpful advice for doing things better or more easily.

snue commented on vmicore/rust_src/.gitignore:
```diff
@@ -1,2 +1,4 @@
 /target/
-Cargo.lock
+# For reproducible builds with Nix, we must check the file in, although it is a
+# library.
```
It's not a "library". Maybe better wording:

```suggestion
# For reproducible builds with Nix, we must check the file in, although it is auto-generated.
```

snue replied:
Ah, I get it now. This refers to the old guidance on committing the lock file for applications but not libraries. FYI: the guidance was updated a while back to "do what is best for your project", see https://blog.rust-lang.org/2023/08/29/committing-lockfiles/

snue reviewed (approved):
LGTM. Left two minor wording suggestions that you can address or ignore.

lbeierlieb committed:
cargo: pin cxx version with patch-level accuracy

This change is necessary to enable offline compilation of the
cxxbridge-cmd with Cargo. The offline compilation is a requirement to
build in the Nix sandbox.

lbeierlieb committed:
cargo: track lock file

To enable reproducible builds (with Nix), Cargo.lock needs to be tracked
in version control.

Commit:
Update Corrosion

Corrosion setup breaks due to new rustup toolchain list format
since v1.28.0

Commit:
Make VmiException public

Commit:
ApiTracing: Use public include for VmiException

Commit:
VmiCore: code reformat

Commit:
InMemory: Code reformat

Commit:
Template: Code reformat

lbeierlieb committed:
fixup! cargo: track lock file

lbeierlieb committed:
cmake: check if correct libvmi is supplied

The fetchContent way of retrieving libvmi specifies the exact branch to
retrieve; with find_package, it is possible that an incorrect version of
libvmi is passed from vcpkg/nix/etc. This commit adds a check that
ensure the functions GDATA added to their libvmi fork are present. The
build would fail later, but this explicit error clarifies the issue.

snue commented on vmicore/rust_src/CMakeLists.txt:
```diff
@@ -1,10 +1,13 @@
+    FetchContent_Declare(
+            Corrosion
+            GIT_REPOSITORY https://github.com/corrosion-rs/corrosion.git
+            GIT_TAG stable/v0.5  # Newer versions require cmake 3.22 which is not available on Debian 11
```
I think the comment is outdated. If you bump the version here, it means Debian 11 is no longer supported as a build platform?

snue replied:
Ah, the bump was done upstream, it seems without updating the comment.

lbeierlieb commented:
This PR was just for internal review before the upstream PR (https://github.com/GDATASoftwareAG/smartvmi/pull/154)
